browser/browser.go

```diff
@@ -0,0 +1,97 @@
+package browser;
+
+import (
+	"os";
+	"io";
+  "fmt";
+	"gopkg.in/vmihailenco/msgpack.v2";
+)
+
+type Command struct {
+	Id string `msgpack:"id"`
+	Cmd string `msgpack:"cmd"`
+	Args []string `msgpack:"args"`
+	Data []byte `msgpack:"data"`
+}
+
+func RunCmd (cmd Command, out io.Writer) {
+	encoder := msgpack.NewEncoder(out);
+	if cmd.Id == "" {
+		encoder.Encode(FailedResultSet("", "No Id supplied."));
+		return;
+	}
+	if len(cmd.Args) == 0 {
+		res := FailedResultSet(cmd.Id, "Not enough arguments.");
+		WriteOut(encoder, res);
+		return;
+	}
+
+	id := cmd.Id;
+
+	switch cmd.Cmd{
```

Comment:
As pete suggests...

Instead of doing a `switch cmd.Cmd {`

how about:
```go
var mCommandRegistry = sync.Mutex{}
var commandRegistry = map[string]func(Command, chan<- interface{})

func RunCmd(cmd Command, outChan chan<- interface{}) {
  mCommandRegistry.Lock()
  handler, ok := commandRegistry[cmd.Cmd]
  mCommandRegistry.Unlock()
  if !ok {
    // Ignore unknown command
  }
  handler(cmd, outChan)
}

func RegisterHandler(cmd string, handler func(Command, chan<- interface)) {
    mCommandRegistry.Lock()
    defer mCommandRegistry.Unlock()
    _, ok := commandRegistry[cmd]
    if (ok) {
        panic('command handler already registered')
    }
    commandRegistry[cmd] = handler
}
```

Then in an other file you can have:
```go
func init() {
  RegisterHandler("ls", list)
}

def list(cmd Command, outChan chan<- interface{}) {
  ...
}

```